File: stats/cloud/aws.go
/* Acksin Autotune - Linux Diagnostics
 * Copyright (C) 2016 Acksin <hey@acksin.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package cloud

import (
	"reflect"

	"github.com/aws/aws-sdk-go/aws/ec2metadata"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"

	"github.com/acksin/autotune/shared"
)

// AWSStats returns the relevant AWS information about the current
// instance via the machine's EC2 Metadata IP interface.
type AWSStats struct {
	AmiID           string `metadata:"ami-id" `
	AmiLaunchIndex  string `metadata:"ami-launch-index"`
	AmiManifestPath string `metadata:"ami-manifest-path"`
	Hostname        string `metadata:"hostname"`
	InstanceAction  string `metadata:"instance-action"`
	InstanceID      string `metadata:"instance-id"`
	InstanceType    string `metadata:"instance-type"`
	LocalHostname   string `metadata:"local-hostname"`
	LocalIpv4       string `metadata:"local-ipv4"`
	MAC             string `metadata:"mac"`
	Profile         string `metadata:"profile"`
	PublicHostname  string `metadata:"public-hostname"`
	PublicIpv4      string `metadata:"public-ipv4"`
	ReservationID   string `metadata:"reservation-id"`
	SecurityGroups  string `metadata:"security-groups"`

	Placement struct {
		AvailabilityZone string `metadata:"availability-zone"`
	} `metadata:"placement"`

	Spot struct {
		Termination string `metadata:"termination" json:",omitempty"`
	} `metadata:"spot" json:",omitempty"`

	// TODO: These need to be converted into iteration tasks.
	// block-device-mapping/
	// metrics/
	// network/
	// public-keys/
	// services/

	// Full
}

func (a *AWSStats) parseMetadata(m *ec2metadata.EC2Metadata) {
	st := reflect.TypeOf(*a)

	for i := 0; i < st.NumField(); i++ {
		metadataTag := st.Field(i).Tag.Get("metadata")
		if metadataTag != "" {
			data, err := m.GetMetadata(metadataTag)
			if err != nil {
				continue
			}

			reflect.ValueOf(a).Elem().Field(i).SetString(data)
		}
	}
}

func (i *AWSStats) getFull() {
	svc := ec2.New(session.New(), &aws.Config{
		Region: aws.String("us-west-2"),
	})

	resp, err := svc.DescribeInstances(nil)
	if err != nil {
		panic(err)
	}

	fmt.Println("> Number of reservation sets: ", len(resp.Reservations))
	for idx, res := range resp.Reservations {
		fmt.Println("  > Number of instances: ", len(res.Instances))
		for _, inst := range resp.Reservations[idx].Instances {
			fmt.Println("    - Instance ID: ", *inst.InstanceId)
		}
	}
}

// NewAWS returns an AWSStats if the current machine is an AWS
// instance otherwise it returns nil.
func NewAWS(a *shared.Config) (i *AWSStats) {
	metadata := ec2metadata.New(session.New())

	// Verify that this is in fact an AWS machine.
	if !metadata.Available() {
		return nil
	}

	i = &AWSStats{}
	i.parseMetadata(metadata)

	if data, err := metadata.GetMetadata("placement/availability-zone"); err == nil {
		i.Placement.AvailabilityZone = data
	}

	if data, err := metadata.GetMetadata("spot/termination"); err == nil {
		i.Spot.Termination = data
	}

	// if a.Cloud != nil && a.Cloud.AWS != nil {
	// 	i.getFull()
	// }

	return
}

File: ai/mental_models/cloud/aws.py
# Copyright (C) 2016 Acksin <hey@acksin.com>
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import json
import os
import boto3

class CloudAWS(object):
    def __init__(self, config, id, aws_access_key, aws_secret_key):
        self.config = config

        self.ID = id
        self.aws_access_key = aws_access_key
        self.aws_secret_key = aws_secret_key

        self.ec2 = boto3.client(
            'ec2',
            aws_access_key_id=aws_access_key,
            aws_secret_access_key=aws_secret_key,
        )

        self.s3 = boto3.client('s3')

    def write_object(self, key, value):
        self.s3.put_object(
            Bucket=self.config['s3_bucket'],
            Key=os.path.join(self.ID, 'ec2', key),
            Body=value.encode('utf-8')
        )

    def get_instances(self):
        desc_instances = self.ec2.describe_instances(Filters=[{'Name': 'instance-state-name', 'Values': ['running']}])
        instances = []

        for reservation in desc_instances[u'Reservations']:
            instances += reservation['Instances']

        for i in instances:
            print i['InstanceId']
            CloudAWSEC2Instance(self, i)

class CloudAWSEC2Instance(object):
    def __init__(self, aws, instance):
        aws.write_object(
            "{}/stats.json".format(instance['InstanceId']),
            json.dumps(instance, cls=AcksinEncoder)
        )

        # CloudAWSMetrics(aws, instance)
        CloudAWSSecurity(aws, instance)

class CloudAWSSecurity(object):
    def __init__(self, aws, instance):
        self.aws = aws
        self.instance = instance

        self.security = []

        self.is_open_to_the_web()

    def is_open_to_the_web(self):
        print self.instance['PublicDnsName']


from datetime import datetime
import decimal


class AcksinEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, decimal.Decimal):
            if o % 1 > 0:
                return float(o)
            else:
                return int(o)
        elif isinstance(o, datetime):
            serial = o.isoformat()
            return serial

        raise TypeError ("Type not serializable")
